src/main/scala/devices/debug/SBA.scala

```diff
@@ -0,0 +1,320 @@
+
+    cover(SBCSFieldsReg.sbautoincrement && SBCSFieldsReg.sbbusy,  "SBCS Cover", "Access with autoincrement set")
+    cover(!SBCSFieldsReg.sbautoincrement && SBCSFieldsReg.sbbusy, "SBCS Cover", "Access without autoincrement set")
+
+    (sbcsfields, sbaddrfields, sbdatafields)
+  }
+}
+
+class SBToTL(implicit p: Parameters) extends LazyModule {
+
+  val cfg = p(DebugModuleParams)
+
+  val node = TLClientNode(Seq(TLClientPortParameters(Seq(TLClientParameters("debug")))))
+
+  lazy val module = new LazyModuleImp(this) {
+    val io = IO(new Bundle {
+      val rdEn         = Bool(INPUT)
+      val wrEn         = Bool(INPUT)
+      val addrIn       = UInt(INPUT, 128) // TODO: Parameterize these widths
+      val dataIn       = UInt(INPUT, 128)
+      val sizeIn       = UInt(INPUT, 3)
+      val dmactive     = Bool(INPUT)
+      val rdLegal      = Bool(OUTPUT)
+      val wrLegal      = Bool(OUTPUT)
+      val rdDone       = Bool(OUTPUT)
+      val wrDone       = Bool(OUTPUT)
+      val respError    = Bool(OUTPUT)
+      val dataOut      = UInt(OUTPUT, 8)
+      val rdLoad       = Vec(cfg.maxSupportedSBAccess/8, Bool()).asOutput
+      val sbStateOut   = UInt(OUTPUT, log2Ceil(SystemBusAccessState.maxId))
+    })
+
+    import SystemBusAccessState._
+ 
+    val (tl, edge) = node.out(0)
+
+    // --- Drive payloads on bus to TileLink ---
+    val muxedData = Wire(init = 0.U(8.W))
+    val requestValid = tl.a.valid
+    val requestReady = tl.a.ready
+    val responseValid = tl.d.valid
+    val responseReady = tl.d.ready
+    val (rdLegal,  gbits) = edge.Get(0.U, io.addrIn, io.sizeIn)
+    val (wrLegal, pfbits) = edge.Put(0.U, io.addrIn, io.sizeIn, muxedData)
+    val (      _,  nbits) = edge.Put(0.U, io.addrIn, io.sizeIn, data=0.U, mask=0.U)
+
+    io.rdLegal := rdLegal
+    io.wrLegal := wrLegal
+
+    val sbState = Reg(init = 0.U)
+    io.sbStateOut := sbState
+    when      (sbState === SBReadRequest.id.U) { tl.a.bits :=  gbits  }
+    .elsewhen (sbState === SBWriteRequest.id.U){ tl.a.bits := pfbits  }
+    .otherwise                                 { tl.a.bits :=  nbits  }
+
+    val respError = tl.d.bits.error
+    io.respError := respError
+
+    val counter = Reg(init = 0.U((log2Ceil(cfg.maxSupportedSBAccess/8)+1).W))
+    val vecData   = Wire(Vec(cfg.maxSupportedSBAccess/8, UInt(8.W)))
+    vecData := Vec.tabulate(16) { i => io.dataIn(8*i+7,8*i) }
+    muxedData := vecData(counter)
+
+    val wrTxValid = sbState === SBWriteRequest.id.U && requestValid && requestReady
+    val rdTxValid = sbState === SBReadResponse.id.U && responseValid && responseReady
+    val txLast    = counter === ((1.U << io.sizeIn) - 1.U)
+    counter := Mux((wrTxValid || rdTxValid) && txLast, 0.U,
+               Mux((wrTxValid || rdTxValid)          , counter+1.U, counter))
+
+    for (i <- 0 until (cfg.maxSupportedSBAccess/8)) {
+      io.rdLoad(i) := rdTxValid && (counter === i.U)
+    }
+
+    // --- State Machine to interface with TileLink ---
```

Comment:
There is a problem: d.ready is low based on a state machine that expects a.fire => can lead to dead-lock. Otherwise, just a few stylistic things I'd change, but not worth getting into here.